relay/relay.go

```diff
@@ -263,261 +242,180 @@ func (r *Relay) process(parentCtx context.Context) error {
-			if r.cfg.EnableGTID {
-				_, currentGtidSet := r.meta.GTID()
-				if currentGtidSet.Equal(r.gSetWhenSwitch) {
-					addRelayFlag = true
-					log.Infof("[relay] binlog events in the gap for file %s already exists in previous sub directory, adding a LOG_EVENT_RELAY_LOG_F flag to them. current GTID set %s", r.fd.Name(), currentGtidSet)
-				} else if currentGtidSet.Contain(r.gSetWhenSwitch) {
-					// only after the first gap filled, currentGtidSet can be the superset of gSetWhenSwitch
-					log.Infof("[relay] binlog events in the gap for file %s not exists in previous sub directory, no need to add a LOG_EVENT_RELAY_LOG_F flag. current GTID set %s, previous sub directory end GTID set %s", r.fd.Name(), currentGtidSet, r.gSetWhenSwitch)
-				} else {
-					return errors.Errorf("currnet GTID set %s is not equal or superset of previous sub directory end GTID set %s", currentGtidSet, r.gSetWhenSwitch)
-				}
+				return errors.Annotate(err, "save meta after recovered")
 			}
+		} else if result.LatestPos.Compare(latestPos) > 0 ||
+			(result.LatestGTIDs != nil && result.LatestGTIDs.Contain(latestGTID)) {
+			log.Warnf("[relay] relay log file have more events after position %s (until %s), GTID sets %v (until %v)",
+				latestPos, result.LatestPos, latestGTID, result.LatestGTIDs)
 		}
 
-		ctx, cancel := context.WithTimeout(parentCtx, eventTimeout)
+	}
+	return errors.Annotatef(err, "recover for UUID %s with config %+v", uuid, cfg)
+}
+
+// handleEvents handles binlog events, including:
+//   1. read events from upstream
+//   2. transform events
+//   3. write events into relay log files
+//   4. update metadata if needed
+func (r *Relay) handleEvents(ctx context.Context, reader2 reader.Reader, transformer2 transformer.Transformer, writer2 writer.Writer) error {
+	var (
+		_, lastPos  = r.meta.Pos()
+		_, lastGTID = r.meta.GTID()
+	)
+
+	for {
+		// 1. reader events from upstream server
+		ctx2, cancel2 := context.WithTimeout(ctx, eventTimeout)
 		readTimer := time.Now()
-		var e *replication.BinlogEvent
-		if gapReader != nil {
-			e, err = gapReader.GetEvent(ctx)
-		} else {
-			e, err = r.reader.GetEvent(ctx)
-		}
-		cancel()
+		rResult, err := reader2.GetEvent(ctx2)
+		cancel2()
 		binlogReadDurationHistogram.Observe(time.Since(readTimer).Seconds())
 
 		if err != nil {
-			switch errors.Cause(err) {
-			case context.Canceled:
+			if rResult.ErrIgnorable {
+				log.Infof("[relay] get ignorable error %v when reading binlog event", err)
 				return nil
-			case context.DeadlineExceeded:
-				log.Infof("[relay] deadline %s exceeded, no binlog event received", eventTimeout)
+			} else if rResult.ErrRetryable {
+				log.Infof("[relay] get retryable error %v when reading binlog event", err)
 				continue
+			}
+			switch err {
 			case replication.ErrChecksumMismatch:
 				relayLogDataCorruptionCounter.Inc()
 			case replication.ErrSyncClosed, replication.ErrNeedSyncAgain:
-				// do nothing
+				// do nothing, but the error will be returned
```

Comment:
also stat it into metric?

Reply:
it will be recorded in `relayExitWithErrorCounter`, should we need to add another metric like `relayGoMySQLError`?